core/src/gateway_createfromjson.c

```diff
@@ -129,49 +131,72 @@ GATEWAY_HANDLE Gateway_CreateFromJson(const char* file_path)
     return gw;
 }
 
+static void rollbackModules(size_t success_modules_entries_count, GATEWAY_HANDLE gw, VECTOR_HANDLE modules_added_successfully)
+{
+    if (success_modules_entries_count > 0)
+    {
+        //Continue adding modules until all are added or one fails
+        for (size_t properties_index = 0; properties_index < success_modules_entries_count; ++properties_index)
+        {
+            GATEWAY_MODULES_ENTRY* entry = (GATEWAY_MODULES_ENTRY*)VECTOR_element(modules_added_successfully, properties_index);
+            if (Gateway_RemoveModuleByName(gw, entry->module_name) != 0)
+            {
+                LogError("Failed to remove module %s up failure.", entry->module_name);
+            }
+        }
+    }
+}
 
 int Gateway_UpdateFromJson(GATEWAY_HANDLE gw, const char* json_content)
 {
     int result;
+    /* Codes_SRS_GATEWAY_JSON_04_004: [ If gw is NULL the function shall return NULL. ] */
     if (gw == NULL)
     {
         LogError("Gw handle can't be NULL.");
         result = __LINE__;
```

Comment:
I mean changing the return type of the function would be a breaking change to the API, and to a lesser degree changing the actual values returned by the function can potentially be a breaking change.